Enable batch support for `windowed_mean|variance`

nicolaspi opened the pull request:
This PR makes functions `windowed_mean` and `windowed_variance` to accept indices with batch dimensions.

Example:

```python
x = np.array([[1, 2, 3], [1, 2, 3], [1, 2, 3]], dtype=np.float32)
low_indices = [[0, 0, 0], [1, 0, 0], [2, 2, 0]]
high_indices = [[3, 3, 3], [1, 2, 3], [3, 2, 1]]
tfp.stats.windowed_mean(x, low_indices=low_indices, high_indices=high_indices, axis=1)
```
Now gives:
```
<tf.Tensor: shape=(3, 3), dtype=float32, numpy=
array([[2. , 2. , 2. ],
       [0. , 1.5, 2. ],
       [3. , 0. , 1. ]], dtype=float32)>
```
Was previously failing with:
```
tensorflow.python.framework.errors_impl.InvalidArgumentError: required broadcastable shapes [Op:SelectV2]
```

nicolaspi commented:
@axch I made changes in the code you authored, could you kindly have a look at this PR?
Thanks

axch commented:
@nicolaspi thanks for the contribution!  I am no longer an active maintainer of TFP, so I'm not really in a position to review your PR in detail (@jburnim please suggest someone?).  On a quick look, though, I see a couple potential code style issues:
- Do we need the dependency on tf.experimental.numpy?
- Do we need the special case for rank-1 indices?  Could we define a more uniform behavior instead?
- I'm guessing some of the shape munging already has relevant helpers defined elsewhere in TFP, but I don't remember off-hand
- TFP generally tries to handle static and dynamic TF shapes uniformly using [prefer_static](https://github.com/tensorflow/probability/blob/main/tensorflow_probability/python/internal/prefer_static.py) and [tensorshape_util](https://github.com/tensorflow/probability/blob/main/tensorflow_probability/python/internal/tensorshape_util.py).

nicolaspi commented:
Thanks for your feedback!

> * Do we need the dependency on tf.experimental.numpy?

We need specifically the `take_along_axis` function that allow to gather the slices along each batch dimensions. I replaced the 'experimental' import path with `from tensorflow.python.ops import numpy_ops`.

> * Do we need the special case for rank-1 indices?  Could we define a more uniform behavior instead?

There is two motivations for this case. First, for backward compatibility, it is equivalent to the legacy non batched usage. Second, it is the only case I can think of where the broadcast is unambiguous when `rank(indices) < rank(x)`.

> * I'm guessing some of the shape munging already has relevant helpers defined elsewhere in TFP, but I don't remember off-hand

In any case, I modified the unit tests to test against non static shapes.

> * TFP generally tries to handle static and dynamic TF shapes uniformly using [prefer_static](https://github.com/tensorflow/probability/blob/main/tensorflow_probability/python/internal/prefer_static.py) and [tensorshape_util](https://github.com/tensorflow/probability/blob/main/tensorflow_probability/python/internal/tensorshape_util.py).

I made usage of `prefer_static` whenever possible.

nicolaspi commented:
@jburnim can you please suggest a reviewer?
CC @axch

SiegeLordEx commented:
I'll take a look at this.

SiegeLordEx commented on tensorflow_probability/python/stats/sample_stats.py:
```diff
@@ -906,30 +925,28 @@ def _prepare_window_args(x, low_indices=None, high_indices=None, axis=0):
-  safe_totals = tf.where(~tf.equal(counts, 0), totals, 0)
-  return tf.where(~tf.equal(counts, 0), safe_totals / counts, 0)
+  indices_shape = ps.shape(low_indices)
+  if ps.rank(low_indices) < ps.rank(x):
```
We can't write this like this. prefer_static will return Tensors if the inputs are not statically known, which cannot appear in a conditional like this. This code needs to be written without the conditional, or, as a last resort, using `tf.cond`.

Alternatively, we could require that the ranks are known, but then you need to do something like:

```
indices_rank = tf.get_static_value(ps.rank(low_indices))
if indices_rank is None:
  raise ValueError(...)
```

nicolaspi replied:
I added the rank check.

SiegeLordEx commented on tensorflow_probability/python/stats/sample_stats_test.py:
```diff
@@ -683,6 +683,129 @@ def test_windowed_mean_corner_cases(self):
+  def test_windowed_mean(self):
+    self.check_windowed(func=tfp.stats.windowed_mean, numpy_func=np.mean)
+
+  def test_windowed_mean_graph(self):
```
There's no need to explicitly test this, `test_all_tf_execution_regimes` decorator already runs each test case in graph mode.

nicolaspi replied:
I removed those tests.

SiegeLordEx commented on tensorflow_probability/python/stats/sample_stats_test.py:
```diff
@@ -683,6 +683,129 @@ def test_windowed_mean_corner_cases(self):
+    check_fn((64, 4, 8), (1, 4), axis=2)
+    check_fn((64, 4, 8), (64, 4), axis=2)
+
+    with self.assertRaises(Exception):
```
This should be assertRaisesRegexp with a more specific error class and a snippet of the error message.

nicolaspi replied:
Done

SiegeLordEx commented on tensorflow_probability/python/stats/sample_stats_test.py:
```diff
@@ -683,6 +683,129 @@ def test_windowed_mean_corner_cases(self):
+  def check_windowed(self, func, numpy_func):
+    check_fn = functools.partial(self.check_gaussian_windowed,
+                                 window_func=func, np_func=numpy_func)
+    check_fn((64, 4, 8), (128, 1, 1), axis=0)
```
This needs to be written as parameterized tests, see this [example](https://github.com/tensorflow/probability/blob/f7da474f215ec33c53265e09882c89db84eb7bbf/tensorflow_probability/python/bijectors/masked_autoregressive_test.py#L993).

nicolaspi replied:
Done

SiegeLordEx commented on tensorflow_probability/python/stats/sample_stats_test.py:
```diff
@@ -683,6 +683,129 @@ def test_windowed_mean_corner_cases(self):
+@test_util.test_all_tf_execution_regimes
+class WindowedStatsTest(test_util.TestCase):
+
+  def _maybe_expand_dims_to_make_broadcastable(self, x, shape, axis):
```
These two functions are as complex as the thing we're testing. Is there any way we can write this via np.vectorize?

nicolaspi replied:
I refactored using np.vectorize, but I am not sure it is easier to read.

nicolaspi commented on tensorflow_probability/python/stats/sample_stats.py:
```diff
@@ -14,10 +14,20 @@
 import numpy as np
 import tensorflow.compat.v2 as tf
 
+if NUMPY_MODE:
```
Thanks for the review!

1. I don't feel comfortable rewriting `take_along_axis` as it would duplicate already existing logics, I feel like it would produce unnecessary maintenance burden.
2. What about [mapping](https://github.com/tensorflow/probability/pull/1600/commits/26f4f121a2ac4b2b14f15af38ebce944ba3f0fa8) `tensorflow.experimental.numpy` to `numpy` and `jax.numpy` backends?

nicolaspi commented on tensorflow_probability/python/stats/sample_stats.py:
```diff
@@ -726,11 +736,17 @@ def windowed_variance(
+  Suppose `x` has shape `Bx + [N] + E`, `low_indices` and `high_indices`
+  have shape `Bi + [M] + F`, such that `rank(Bx) = rank(Bi) = axis`.
+  Then each element of `low_indices` and `high_indices` must be
+  between 0 and N+1, and the shape of the output will be `Bx + [M] + E`.
```
I updated to `broadcast(Bx, Bi) + [M] + broadcast(E, F)`

nicolaspi commented on tensorflow_probability/python/stats/sample_stats.py:
```diff
@@ -726,11 +736,17 @@ def windowed_variance(
-  must be between 0 and N+1, and the shape of the output will be
-  `Bx + [M] + E`.  Batch shape in the indices is not currently supported.
+  Suppose `x` has shape `Bx + [N] + E`, `low_indices` and `high_indices`
+  have shape `Bi + [M] + F`, such that `rank(Bx) = rank(Bi) = axis`.
```
Please check my comment below.

nicolaspi commented on tensorflow_probability/python/stats/sample_stats.py:
```diff
@@ -726,11 +736,17 @@ def windowed_variance(
+
+  The shape `Bi + [1] + F` must be broadcastable with the shape of `x`.
+
+  If `rank(Bi + [M] + F) < rank(x)`, then the indices are expanded
```
We specify the implicit rules we uses for broadcasting. I updated the formulation.

nicolaspi commented on tensorflow_probability/python/stats/sample_stats.py:
```diff
@@ -726,11 +736,17 @@ def windowed_variance(
+  Then each element of `low_indices` and `high_indices` must be
+  between 0 and N+1, and the shape of the output will be `Bx + [M] + E`.
+
+  The shape `Bi + [1] + F` must be broadcastable with the shape of `x`.
```
> This contradicts the next paragraph, no?

Yes, I reformulated.

> The batching would introduce a batch dimension on the left of those shapes: 
> Thus, the only broadcasting requirements are that Bx and Bi broadcast. I don't know where F came from.

Maybe the term 'batch' is not proper. This contribution adds the possibility to have the more general case where 
idx shape is `Bi + [M] + F`. F could be seen as 'inner batch dimensions', but here 'batch' carries a different semantic than the standard machine learning one where it is represented by outer dims.
